[ws-proxy] Add ws-manager TCP proxy

corneliusludmann opened the pull request:
This is my proposed solution for a TCP proxy from `ws-proxy`  to `ws-manager`. Open for discussion.

Test it with:
```
$ ./gpctl workspaces --pod ws-proxy --port 8081 list
```

Issue: #3585

corneliusludmann commented on components/ws-proxy/pkg/utils/rate-limited-listener.go:
```diff
@@ -0,0 +1,80 @@
+	if err != nil {
+		return
+	}
+	log.WithField("remote addr", conn.RemoteAddr()).Debug("RateLimitedListener: new connection incoming")
```
Note to myself: Remove debug log message.

corneliusludmann commented on components/ws-proxy/pkg/utils/rate-limited-listener.go:
```diff
@@ -0,0 +1,80 @@
+	log.WithField("remote addr", conn.RemoteAddr()).Debug("RateLimitedListener: new connection incoming")
+	err = rll.ratelimiter.Wait(rll.ctx)
+	if err != nil {
+		log.
+			WithError(err).
+			WithField("remote addr", conn.RemoteAddr()).
+			Info("RateLimitedListener: error from ratelimiter")
```
Note to myself: Remove debug log message.

csweichel commented on components/ws-proxy/cmd/root.go:
```diff
@@ -114,3 +150,30 @@ func getConfig(fn string) (*Config, error) {
+
+// Duration is a json-unmarshallable wrapper for time.Duration.
+// See https://stackoverflow.com/questions/48050945/how-to-unmarshal-json-into-durations
+type Duration time.Duration
+
+// UnmarshalJSON unmarshales a duration using ParseDuration
+func (d *Duration) UnmarshalJSON(b []byte) error {
+	var v interface{}
+	if err := json.Unmarshal(b, &v); err != nil {
+		return err
+	}
+
+	switch value := v.(type) {
+	case float64:
+		*d = Duration(time.Duration(value))
+		return nil
+	case string:
+		tmp, err := time.ParseDuration(value)
+		if err != nil {
+			return err
+		}
+		*d = Duration(tmp)
+		return nil
+	default:
+		return xerrors.New("invalid duration")
+	}
+}
```
this already exists in https://github.com/gitpod-io/gitpod/blob/b7dda84ad0c0dd195b829b7f4160f9f7acfa355c/components/common-go/util/duration.go#L14-L48

We should be able to re-use that bit, rather than copy.

csweichel commented on components/ws-proxy/cmd/run.go:
```diff
@@ -93,11 +96,26 @@ var runCmd = &cobra.Command{
 		}
 
+		ctx, cancel := context.WithCancel(context.Background())
+		if len(cfg.WSManagerProxy.ListenAddress) > 0 {
```
```suggestion
		if cfg.WSManagerProxy.ListenAddress != "" {
```

csweichel commented on components/ws-proxy/pkg/utils/rate-limited-listener.go:
```diff
@@ -0,0 +1,80 @@
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package utils
```
nit: `utils` is an actively discouraged package name (see [here](https://blog.golang.org/package-names), **Avoid meaningless package names**) ... maybe `ratelimit`

corneliusludmann replied:
Good point. Shame on me for being too lazy to find a proper package name. :laughing:

corneliusludmann commented:
@csweichel Thanks for your comments which I addressed with the latest push.

I also added the possibility to disable rate limiting by config (just in case):

https://github.com/gitpod-io/gitpod/blob/a1fa5ef3a6bd96d474230666cc2a2210bf24439c/components/ws-proxy/cmd/run.go#L133-L135

csweichel commented on components/ws-proxy/pkg/ratelimit/rate-limited-listener.go:
```diff
@@ -0,0 +1,78 @@
+// NewRateLimitedListen returns a net.Listen implementation that adds rate limiting.
+func NewRateLimitedListen(ctx context.Context, refillInterval time.Duration, bucketSize int) func(network, address string) (net.Listener, error) {
+	return func(network, address string) (net.Listener, error) {
+		return newRateLimitedListener(
+			ctx,
+			network,
+			address,
+			refillInterval,
+			bucketSize,
+		)
```
nit: why the indirection? Couldn't the code in `newRateLimitedListener` be here directly?

corneliusludmann replied:
Right. It's a left-over from code restructuring.

csweichel commented on chart/templates/ws-proxy-configmap.yaml:
```diff
@@ -60,6 +60,15 @@ data:
                 "location": "/app/public"
             }
         },
+        {{ if (and $comp.wsManagerProxy $comp.wsManagerProxy.enabled) }}
```
bonus feature: the helm chart could fail if someone were to expose their ws-manager but had no auth/TLS configured.

corneliusludmann replied:
Good idea. But we don't have TLS config in ws-manager template yet:

https://github.com/gitpod-io/gitpod/blob/555ca137b0f68d02caa1b18248b46f34619b5fd5/chart/templates/ws-manager-configmap.yaml#L95-L111

Already working on a follow-up PR to add this.

csweichel reviewed (approved):
LGTM.

I have two more nits/bonus feature things, that could also be addressed in a follow up
